ipq40xx: convert Engenius EAP1300 to DSA

cminyard opened the pull request:
This commit converts the EAP1300 to DSA setup.

Pretty straightforward to do.  The hardest part was finding the location of the MAC address.

I have tested this on main and on 23.05.3 (though the location of  qcom-ipq4018-eap1300.dts is different in the two versions).

If we could get this included in mainstream and 23.05.3, that would be great.

Thank you.

cminyard committed:
ipq40xx: convert Engenius EAP1300 to DSA

This commit converts the EAP1300 to DSA setup.

Signed-off-by: Corey Minyard <minyard@acm.org>
Link: https://github.com/openwrt/openwrt/pull/15358
Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

robimarko commented on target/linux/ipq40xx/files-6.6/arch/arm/boot/dts/qcom/qcom-ipq4018-eap1300.dts:
```diff
@@ -174,6 +174,13 @@
+				#address-cells = <1>;
+				#size-cells = <1>;
+
+				macaddr_orgdata_71f28: macaddr@71f28 {
```
Are you sure about this?

Because, it makes no sense that MAC-s are stored in the U-Boot partition

cminyard replied:
It is a bit weird.  I searched through all the memory and it's the only place I found with the MAC address, though.  I'll look again.  I realized that I hadn't verified the MAC addresses of the WIFI devices, so I need to check that anyway.

cminyard replied:
You are right.  I booted the factory firmware, and the MACs are not what the 22.03.5 version was showing.  There is a MAC address reported by the u-boot, which is what I was using, but that is apparently fixed in the firmware and wrong.

The 22.03.5 version comes up with a MAC address on the ethernet devices that is whatever is left over in the device, which appears to be fixed.  The wifi macs come from 6 bytes into the configuration data, which appears to be some kind of default, but again is wrong.

The only place I could find the MAC address for eth0 is the u-boot-env partition, there is a macaddr= field there that is correct.  The wifi MAC addresses are the macaddr value incremented by 1 and 2, respectively.  There is a wanaddr= field that is what the eth device comes up as if you don't set anything, so I'm assuming u-boot is setting that.

u-boot is not setting the MAC in the device tree like it normally would.  I'm guessing in the OEM load some software comes up later, read the MAC address from the u-boot environment, sets the eth device, and generates the others for the wifi devices.  What can I do here?

Ansuel replied:
Check other dts, uboot env entry can be parsed as nvmem cell and used to reference mac addr... then with the help of nvmem-layout you can increment it by 1 and 2 for wifi. It done my times in other dts so you should be able to find many examples

keyword are nvmem-layout and uboot or uboot,env if i'm not wrong.

robimarko replied:
https://github.com/openwrt/openwrt/blob/main/target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-wax218.dts#L174

One example, but do note that this only works if the U-Boot variable name is ethaddr or ethNaddr for multiple addresses

cminyard committed:
ipq40xx: Enable NVMEM_U_BOOT_ENV config

It's needed to get the MAC addresses for the Engenius EAP1300.

Signed-off-by: Corey Minyard <minyard@acm.org>
Link: https://github.com/openwrt/openwrt/pull/15358
Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

cminyard commented:
Ok, I think I have everything working properly.

I also have a 23.05 version of this, so I submit that as a separate request?  The changes are pretty close, I think, mostly the files go in different places.

And, it looks like someone attempted this a while ago.  See https://github.com/openwrt/openwrt/pull/12209.

DragonBluep commented on target/linux/ipq40xx/files-6.6/arch/arm/boot/dts/qcom/qcom-ipq4018-eap1300.dts:
```diff
@@ -221,16 +226,36 @@
-	nvmem-cells = <&precal_art_5000>;
+	nvmem-cell-names = "pre-calibration", "mac-address";
+	nvmem-cells = <&precal_art_5000>, <&macaddr_ubootenv_ethaddr 2>;
+	mac-address-increment = <2>;
```
`mac-address-increment` is deprecate please remove it.

cminyard replied:
Well I thought I had removed it.  I missed this one.  It doesn't do anything in 6.6, anyway.

cminyard commented:
One more try, with the bogus line removed...

cminyard committed:
ipq40xx: Properly set MAC addresses for the EAP1300

The code that was there was just taking whatever was left in the
registers, which was just wrong.  Set the addresses using the value from
the u-boot environment, the same way the OEM firmware does.

Signed-off-by: Corey Minyard <minyard@acm.org>
Link: https://github.com/openwrt/openwrt/pull/15358
Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

cminyard committed:
ipq40xx: Enable the Engenius EAP1300

Now that DSA is enabled and the MAC addresses are set properly, we can
use it.

Signed-off-by: Corey Minyard <minyard@acm.org>
Link: https://github.com/openwrt/openwrt/pull/15358
Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

cminyard commented:
I expect the next issue is that a single git commit did two things, so I've split it out the MAC fix and the DSA enabling into two patches, and then the final system enabling at the end.

ryancdotorg commented:
+1 to pulling this into the the next 23.x release

cminyard commented:
> +1 to pulling this into the the next 23.x release

Is this something automatic or should I do a pull request?  The patches won't apply as-is, they need some adjustment for file locations.

ryancdotorg commented:
It seems to have been merged? What needed to be changed?

I created another PR here that I believe is required to allow firmware updates to work, and also adds support for the EAP1300EXT as an alt device: https://github.com/openwrt/openwrt/pull/15401

ryancdotorg commented:
git cherry-pick was able to automatically apply the commits against 23.05 without a problem - I've made a PR for that.

@cminyard can you please check #15402?

cminyard commented:
On Mon, May 06, 2024 at 08:05:00AM -0700, Ryan Castellucci wrote:
> It seems to have been merged? What needed to be changed?

It was merged into main, but not in to 23.x.  The config file and the
DTS file are in different locations now, IIRC, but I believe the patches
will apply with that adjusted.

> 
> I created another PR here that I believe is required to allow firmware updates to work, and also adds support for the EAP1300EXT as an alt device: https://github.com/openwrt/openwrt/pull/15401

Yes, that looks good if the eap1300ext is really the same as the
eap1300.

ryancdotorg commented:
I have both an EAP1300 and EAP1300EXT that I've added serial pins to, and I have root on the vendor firmware. Per the commit adding it as an alt device, the vendor firmware is identical for the two devices except the header that says what device it's for. Same kernel, bootloader, root filesystem.
